torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -806,11 +806,13 @@ ProcessGroupNCCL::CUDAEventCache::CUDAEventCache() = default;
 std::shared_ptr<at::cuda::CUDAEvent> ProcessGroupNCCL::CUDAEventCache::create(
     bool timing) {
   // register the deleter as a callback when the WorkNCCL object is destroyed.
-  auto deleter = [this, timing](at::cuda::CUDAEvent* event) {
-    std::lock_guard<std::mutex> lock(this->cacheMutex_);
+  auto deleter = [cache =
+                      std::shared_ptr<ProcessGroupNCCL::CUDAEventCache>(this),
```

Comment:
This may create a separate reference counting than the original one created (by `std::make_shared`). That may be the reason we see double free. We may need to use [`share_from_this`](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this/shared_from_this) instead, to make it one single count.

Reply:
Just writing down my thought process:
each deleter will keep a reference to the `CUDAEventCache`. And the deleter goes with the life of the `CUDAEvent`. That is, there would be (roughly) as many references to the cache as there are `CUDAEvent`s. When the `CUDAEvent` is destructed, the deleter is gone, releasing one ref. After all `CUDAEvent`s being destructed, the ref count drops to 0, then the cache would be destructed -- and so we keep the life of the cache longer than the thread that creates it. The map is still gone as the creating thread dies, but since the map holds only a shared_ptr to the cache, it would not destruct the cache, but only reducing the ref count by 1.

Reply:
Yeah that was what I planned to do originally. I misunderstood how the ref count works here. (I thought using `std::shared_ptr` will reuse the original ref count) Using `[share_from_this](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this/shared_from_this)` indeed works. And now the unit test passed.